linker/Linker.Steps/MarkStep.cs

```diff
@@ -1790,6 +1790,7 @@ protected virtual void MarkRequirementsForInstantiatedTypes (TypeDefinition type
 			Annotations.MarkInstantiated (type);
 
 			MarkInterfaceImplementations (type);
+			MarkMethodIf (type.Methods, m => m.IsFinalizer ());
```

Comment:
Could you fold that with the next MarkMethodsIf. For example `m.IsFinalizer () || IsVirtualAndHasPreservedParent(m)`

Reply:
I didn't because I think there is some chance that future improvements will change the behavior of IsVirtualAndHasPreservedParent and I think keeping it out of there would help reduce the chances of accidentally breaking `Finalize()` in the future.

Also, fundamentally, I don't believe `Finalize()` falls into the same category as `IsVirtualAndHasPreservedParent`.

`Finalize()` is a method required by the runtime (technically the GC?) once an instance of the type could exist.

Where as `IsVirtualAndHasPreservedParent` is a blunt tool to keep things that might be needed by other managed code because we don't have the information available today to know for certain if they are needed.  

I've been thinking about more improvements that can be made and one idea that I have that is relatively trivial and would impact `IsVirtualAndHasPreservedParent` but not `Finalize()` is escape analysis.  If a type never escapes the method(s) it is created in, then we can know all of the calls that were made and we wouldn't need to mark everything that `IsVirtualAndHasPreservedParent` marks today.  I realize it's a little more complicated than that and this may never be worth doing, but I bring it up as an example to highlight why I think `Finalize()` handling shouldn't be bundled with `IsVirtualAndHasPreservedParent`

Reply:
Actually, thinking about this further.  `IsVirtualAndHasPreservedParent` should explicitly not mark `Finalize()`

Just because one types Finalize() is marked or because System.Object.Finalize() is marked does not mean that all types need their Finalize() marked.  Only types that are instantiated need their Finalize() marked.

Reply:
> Actually, thinking about this further.  `IsVirtualAndHasPreservedParent` should explicitly not mark `Finalize()`
> 
> Just because one types Finalize() is marked or because System.Object.Finalize() is marked does not mean that all types need their Finalize() marked.  Only types that are instantiated need their Finalize() marked

Scratch that.  I got mixed up.

Reply:
I got `IsVirtualAndHasPreservedParent` and `ProcessOverrides` mixed up, but my feelings remain the same.  `Finalize()` is a different case than `IsVirtualAndHasPreservedParent` and I don't want to mix the two and risk causing bugs in the future.

Reply:
OK.  I adjusted the code so that we do both in 1 pass over `type.Methods`.  However, I switched to a foreach loop w/ if's.  I don't like Finalize being packed on 1 line with `IsVirtualAndHasPreservedParent`, it hides the behavior in my opinion.